### PR TITLE
A fancy bookseller

### LINKED ISSUE #30: Example: Bookseller
We should have a "maximally fancy" bookseller example, that somehow combines HasChor's best features (higher-order, location polymorphic) and uses multicast, enclaving, and ideally fanout/fanin.

On the other hand, maybe we'd rather just say "Here's how you'd do the stuff HasChor did, in our system from scratch" without trying to make Bookseller into an advanced example...

### PR DESCRIPTION
**Pull request by gshen42**
Resolve #30, still in progress.

This is an attempt to create a maximally fancy bookseller protocol that uses every feature we have. It allows an arbitrary number of *supporters* to contribute money and uses enclaves to constrain the buying decision to only buyer and seller. But I need some help to finish this:

1. How do I complete these proofs?
https://github.com/ShapeOfMatter/MultiChor/blob/767f6e3892e0a8a944e2b1d5dcebf8241461d0a9/examples/BooksellerFancy.hs#L21-L25

2. Does this part of the code make sense? I'm especially curious about if line 43 is necessary.
https://github.com/ShapeOfMatter/MultiChor/blob/767f6e3892e0a8a944e2b1d5dcebf8241461d0a9/examples/BooksellerFancy.hs#L39-L46

3. I have trouble using census polymorphism to finish this. 
https://github.com/ShapeOfMatter/MultiChor/blob/767f6e3892e0a8a944e2b1d5dcebf8241461d0a9/examples/BooksellerFancy.hs#L69-L71

### CODE REVIEW

**Comment by ShapeOfMatter** [comment]
1. off the cuff:
```haskell
proof1 :: Subset ("buyer" ': supporters) ("buyer" ': "seller" ': supporters) 
proof1 = (listedFirst @"buyer") @@ (consSuper . consSuper $ refl @supporters)
 
proof2 :: Subset '["buyer", "seller"] ("buyer" ': "seller" ': supporters) 
proof2 = listedFirst @@ listedSecond @@ nobody
```
There's other options. `listedFirst` is awkward to have around, but turns out to be useful. We could unpack everything so we're not using `Batteries` at all, or we could _probably say
```haskell
proof1 :: Subset ("buyer" ': supporters) ("buyer" ': "seller" ': supporters) 
proof1 = explicitMember @@ (consSuper . consSuper $ refl @supporters)  -- I don't think we can use explicitSubset here...
 
proof2 :: Subset '["buyer", "seller"] ("buyer" ': "seller" ': supporters) 
proof2 = explicitSubset
```
Or we could unpack all the way and write the actual `Subset` functions, but the you're not "supposed" to _need_ to do that.

**Comment by ShapeOfMatter** [comment]
2. Again, an abundance of options. `enclaveTo` is probably the most obviously applicable and the lightest change here. (I don't especially like it; it feels weird supplying an argument that proves something about the _return_ type, but it works) Or just use `flatten` directly. 

I haven't played w/ this specifically, but I don't think there's any reason to use `broadcast'` over `broadcast`?

**Comment by ShapeOfMatter** [comment]
3. Either `parallel` and then `gather`, or `fanIn` containing `locally` and `~>`. Once you have a `Quire`, it's `Foldable`, so `sum` _per se_ will work.

File: examples/BooksellerFancy.hs [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+import CLI
+import Choreography
+import Choreography.Network.Http
+import Data (deliveryDateOf, priceOf)
+import System.Environment
+
+$(mkLoc "buyer")
+$(mkLoc "seller")
+
+proof1 :: Subset ("buyer" ': supporters) ("buyer" ': "seller" ': supporters)
```
Comment (ShapeOfMatter) [inline examples/BooksellerFancy.hs:21]: I think our convention should be that proofs' names reflect their use as _identifiers_, so I'd call these `buyers` (or `customers`?) and maybe `transactors`. Those aren't great names though...

Comment (ShapeOfMatter) [reply examples/BooksellerFancy.hs:21]: Oh, also note that `supporters` here is free; it's a complely different type-variable from the one in `bookseller`. These should probably be declared as `let`s or `where`s in `bookseller` (but I think they will _work_ here).

File: examples/BooksellerFancy.hs [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+  price <- (seller, \un -> return $ priceOf (un seller database) (un seller title)) ~~> buyer @@ nobody
+
+  -- the buyer and supporters make a decision using the `mkDecision` choreography
+  decision <- enclave proof1 $ mkDecision price
+
+  -- get ride of the extra Located (TODO: make a helper function for this?)
+  decision <- locally buyer $ \un -> return (un buyer (un buyer decision))
+
+  -- if the buyer decides to buy the book, the seller sends the delivery date to the buyer
+  enclave proof2 $ broadcast' First (First, decision) >>= \case
```
Comment (ShapeOfMatter) [inline examples/BooksellerFancy.hs:46]: The proofs bound by `mkLoc` should be polymorphic over the super-set; they'll work for any superset where the Symbol appears explicitly.

File: examples/BooksellerFancy.hs [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+    True -> do
+      deliveryDate <- (seller, \un -> return $ deliveryDateOf (un seller database) (un seller title)) ~~> buyer @@ nobody
```
Comment (ShapeOfMatter) [inline examples/BooksellerFancy.hs:48]: There's also `*~>`, which would spare you the `return`. I'm afraid there's _not_ a clear logic to which helper functions exist or what they're called though :(

**Comment by gshen42** [comment]
I’m almost done with this except for one thing: The following code intends to let all supporters input their contributions and send them to the buyer.  I'm using a `fanIn` for this as suggested, but the property test fails. My speculation is that the `fanIn` includes buyer in the loop body and will make buyer send a message to itself, which never happens. Is there a way to exclude `buyer` from the loop body? Or is there something else I'm missing?

https://github.com/ShapeOfMatter/MultiChor/blob/6b1f44473e51921340e33caec1047289f57b67f1/examples/BooksellerFancy.hs#L67-L72

**Comment by ShapeOfMatter** [comment]
I think the git log is going to be ok; IDK exactly what this situation means or how it will shake out, but I was able to force my local machine to show me a clean diff, so it's clearly possible.

**Comment by ShapeOfMatter** [comment]
I see some problems so far in `mkDecision2`:

1. As you've guessed, `"buyer"` is included in the "senders" (`qs`) of the `fanIn`. On its own _that's fine_. What's weird is that we're not getting any kind of error or warning at compile time: AFAICT there's nothing telling GHC what list to use for `qs`, so how is it deciding? In any case, the type arguments of `fanIn` are arranged to make type application easy here.
2. The immediate error is being thrown because `"buyer"` is _also_ providing their budget explicitly (outside the `fanIn`). This is causing a `CLI` error during test evaluation bc we're only giving it the one value on its input tape. Treating `"buyer"` as a contributor or having them enter their budget separately would be fine, it's both that's a problem. 
3. I _think_ the math on line 72 is wrong; the supporter contributions should be added to the budget or subtracted from the price, right?

**Comment by gshen42** [comment]
Thanks! You're right about all the points above. I've fixed the issue and now it passes the test.

BTW, the setup for property-based testing is really nice.

**Comment by ShapeOfMatter** [comment]
Closing bc we manually merged these changes locally.

**Pull request closed without merging**